chain-vote/src/private_voting/unit_vector_zkp.rs

```diff
@@ -0,0 +1,475 @@
+                .iter()
+                .zip(cx.exp_iter())
+                .fold(Scalar::zero(), |acc, (r, cx_pows)| {
+                    let el = r * cx_pows;
+                    el + acc
+                });
+            p1 + p2
+        };
+
+        Proof {
+            ibas: first_announcement_vec,
+            ds: poly_coeff_enc,
+            zwvs: randomness_response_vec,
+            r: response,
+        }
+    }
+
+    /// Verify a unit vector proof. The verifier checks that the plaintexts encrypted in `ciphertexts`,
+    /// under `public_key` is a unit vector.
+    /// Common Reference String (`crs`): Pedersen Commitment Key
+    /// Statement: public key `pk`, and ciphertexts `encrypting_vote.ciphertexts`
+    /// C_0=Enc_pk(r_0; v_0), ..., C_{m-1}=Enc_pk(r_{m-1}; v_{m-1})
+    ///
+    /// Description of the verification procedure available in Figure 9.
+    pub(crate) fn verify(
+        &self,
+        crs: &Crs,
+        public_key: &PublicKey,
+        ciphertexts: &[Ciphertext],
+    ) -> bool {
+        let ck = CommitmentKey::from(crs.clone());
+        let ciphertexts = Ptp::new(ciphertexts.to_vec(), Ciphertext::zero);
+        let bits = ciphertexts.bits();
+        let mut cc = ChallengeContext::new(&ck, public_key, ciphertexts.as_ref());
+        let cy = cc.first_challenge(&self.ibas);
+        let cx = cc.second_challenge(&self.ds);
+
+        if self.ibas.len() != bits {
+            return false;
+        }
+
+        if self.zwvs.len() != bits {
+            return false;
+        }
+
+        self.verify_statements(public_key, &ck, &ciphertexts, &cx, &cy)
+    }
+
+    /// Final verification of the proof, that we compute in a single vartime multiscalar
+    /// multiplication.
+    #[cfg(feature = "ristretto255")]
+    fn verify_statements(
+        &self,
+        public_key: &PublicKey,
+        commitment_key: &CommitmentKey,
+        ciphertexts: &Ptp<Ciphertext>,
+        challenge_x: &Scalar,
+        challenge_y: &Scalar,
+    ) -> bool {
+        let bits = ciphertexts.bits();
+        let length = ciphertexts.len();
+        let cx_pow = challenge_x.power(bits);
+
+        let powers_cx = challenge_x.exp_iter();
+        let powers_cy = challenge_y.exp_iter();
+
+        let powers_z_iterator = powers_z_encs_iter(&self.zwvs, challenge_x, &(bits as u32));
+
+        let zero = public_key.encrypt_with_r(&Scalar::zero(), &self.r);
+
+        let mega_check = GroupElement::multiscalar_multiplication(
+            self.zwvs
+                .iter()
+                .map(|zwv| zwv.z)
+                .chain(self.zwvs.iter().map(|zwv| zwv.w))
+                .chain(iter::repeat(challenge_x.negate()).take(bits))
+                .chain(iter::repeat(Scalar::one().negate()).take(bits))
+                .chain(self.zwvs.iter().map(|zwv| zwv.v))
+                .chain(self.zwvs.iter().map(|zwv| zwv.z - challenge_x))
+                .chain(iter::repeat(Scalar::one().negate()).take(bits))
```

Comment:
Wondering if it is possible to provide a value that was not the committed one (i.e. first part of the multiscalar mul is not zero), while still getting `GroupElement::zero` in the end of the computation

Reply:
yes, absolutely. Forgot an important detail.

Reply:
so what I wanted to do initially (that I forgot, and it is absolutely crucial that you figured it out, thanks 🙏 ) was to use a random scalar, computed by the verifier to be able to include all checks in a single multiscalar mult (see current code). However, with this new random number I thought the code did not read very well, so I thought of breaking it into `bits` multiscalar multiplications for the first part of the verification. But efficiency was slightly affected, and it did not read much better. So I reverted to the initial idea---everything in a single multiscalar mult. WDYT?

Reply:
Also, with this verification procedure, the verifier needs to compute a random number. Also had a self debate there. Should the verifier directly use `thread_rng`, or should we give as input a `CryptoRng + RngCore`? I chose the former, as otherwise we had to handle RNGs in the mockchain code. So basically, `fn verify_statements()` directly uses `thread_rng()`.